electrum/util.py

```diff
@@ -108,6 +108,16 @@ def base_unit_name_to_decimal_point(unit_name: str) -> int:
     except KeyError:
         raise UnknownBaseUnit(unit_name) from None
 
+def check_max_spend(input) -> bool:
+    return isinstance(input, str) and input[-1] == '!'
+
+def parse_max_spend(input: str) -> Decimal:
+    x = ''
+    if input == '!':
+        x = '1'
+    else:
+        x = input[:-1]
+    return Decimal(x)
```

Comment:
We should restrict the weights to be positive integers only; Decimal is overkill.
Also, `parse_max_spend` currently assumes `check_max_spend(input) is True`; might as well just add an assert for it near the beginning of `parse_max_spend`.
Also, `input` is not a good name for the function parameter here...

However looking at how you are using these two functions, I think it might make more sense to merge them into a single one.
Why don't we do something like this?

```suggestion
def parse_max_spend(amt: Any) -> Optional[int]:
    """Checks if given amount is "spend-max"-like.
    Returns None or the positive integer weight for "max". Never raises.

    When creating invoices and on-chain txs, the user can specify to send "max".
    This is done by setting the amount to '!'. Splitting max between multiple
    tx outputs is also possible, and custom weights (positive ints) can also be used.
    For example, to send 40% of all coins to address1, and 60% to address2:
    ```
    address1, 2!
    address2, 3!
    ```
    """
    if not (isinstance(amt, str) and amt and amt[-1] == '!'):
        return None
    if amt == '!':
        return 1
    x = amt[:-1]
    try:
        x = int(x)
    except ValueError:
        return None
    if x > 0:
        return x
    return None
```
This of course means you would also have to modify the call sites (to adapt to the new "API").